### ISSUE TITLE
Metadata for histograms, __slots__

### ISSUE THREAD

**Issue by henryiii** [scikit-hep/boost-histogram#400]
The histograms should also have a metadata slot that can be carried along during computations. This is needed to attach arbitrary metadata for 2-way ROOT conversions (or for names, etc). This does not need C++ support underneath, like axes do, since it doesn't change when slicing, etc, while axes may be changed or removed in an operation.

The implementation would be simple; metadata would be supported for histograms as a keyword only argument / property. Any time a new histogram is returned, it would shallow copy `metadata = copy.copy(self.metadata)`.

I usually set `__slots__` on classes; this maybe should be set once we have a `.metadata` slot; besides a _tiny_ memory savings, we can ensure that metadata is *only* set in the metadata slot, which is propagated. Subclasses could then either implement `__slots__` or not, if they would rather a `__dict__`. We can also provide documentation in the slots value.

**Comment by HDembinski** [comment]
Why not just have a __dict__?

**Comment by henryiii** [comment]
I may have really overthought this... Yes, we should allow a dict, the question is:

```python
new_hist = h[1:5]
```

Should `new_hist` have the same `__dict__` as `h`, or just metadata gets passed along?

**Comment by henryiii** [comment]
For axes, we need to make sure that all operations pass the metadata along with the object, like

```python
new_hist = h[1:5]
assert new_hist.axes[0].metadata == h.axes[0].metadata
```

However, if we could make this the `__dict__` that gets passed along, this could be very useful. You could set a title / label and have it work in plotting tools that expect them, even if boost-histogram itself does not assign special meaning to those.

**Comment by HDembinski** [comment]
I think slices should have shallow copies of metadata, yes.

**Comment by henryiii** [comment]
Solved in #403
